### PR TITLE
Securing Koa with pipeline 2

### PR DESCRIPTION
**Pull request by brunokrebs**

### CODE REVIEW

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 1]
```diff
@@ -28,15 +28,15 @@ related:
...
 
-Since we want our application to secure ours users' data, we will use JWT tokens to authorize certain requests. A JWT - which stands for JSON Web Token and is pronounced as "jot" - is a token that provides credibility on an end to end communication. JWTs are getting widely adopted, and they take place as an alternative to the, rather old, cookies approach. The biggest advantage of JWTs is that they can hold sensitive data, in a readable format, and be trustworthy while getting sent over the network.
+Since we want our application to secure ours users' data, we will use JWT tokens to authorize certain requests. A JWT—which stands for JSON Web Token and is pronounced as "jot"—is a token that provides credibility on an end to end communication. JWTs are getting widely adopted, and they take place as an alternative to the, rather old, cookies approach. The biggest advantage of JWTs is that they can hold sensitive data, in a readable format, and be trustworthy while getting sent over the network.
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: cookies approach => cookie based approach.

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 2]
```diff
@@ -28,15 +28,15 @@ related:
...
 
-Koa is web framework, just like [Express](http://expressjs.com/), that is developed by the many of the same people that built Express - by the way, here is a [nice tutorial on how to secure an Angular 2 app backed by Express](https://auth0.com/blog/angular-2-authentication/). Unofficially known as Express' successor, Koa uses generators to improve readability and robustness of applications. Writing middlewares to handle users requests become very easy and clear with Koa's approach, as we will see on our own grocery list application.
+Koa is web framework, just like [Express](http://expressjs.com/), that is developed by the many of the same people that built Express—by the way, here is a [nice tutorial on how to secure an Angular 2 app backed by Express](https://auth0.com/blog/angular-2-authentication/). Unofficially known as Express' successor, Koa uses `async` functions to improve readability and robustness of applications. Writing middlewares to handle users requests become very easy and clear with Koa's approach, as we will see on our own grocery list application.
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: the many => many

no need for "the"

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 3]
```diff
@@ -122,30 +122,29 @@ export class Exception extends Error {
 
 As we can see, what we've just created is responsible for carrying two properties related to errors. A status code, that represents an HTTP status, and a message with the error description.
 
-In order to be able to warn our users properly - i.e. in a JSON object format - about this errors, we have now to create an `exception-handler.middleware.ts` file that will contain our's middleware source code. Create it as a sibling to `app.ts`, under `./src/server/`. The contents of this middleware are very simple:
+In order to be able to warn our users properly—i.e. in a JSON object format—about this errors, we have now to create an `exception-handler.middleware.ts` file that will contain our's middleware source code. Create it as a sibling to `app.ts`, under `./src/server/`. The contents of this middleware are very simple:
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: this => these

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 4]
```diff
@@ -266,20 +265,20 @@ import {SINGLETON as UserDAO} from "./user.dao";
...
+        user.items = ctx.request.body.items;
         UserDAO.update(user);
-        this.body = {};
+        ctx.body = {};
     }
 }
 ```
 
-Everything that we do in this file is export a object literal with two properties. The first property, called `path`, represents the URL that will be requested by external resources, like our front-end, to update the a user's grocery list. The second property, called `middleware`, contains the generator that will handle such requests.
+Everything that we do in this file is export a object literal with two properties. The first property, called `path`, represents the URL that will be requested by external resources, like our front-end, to update the a user's grocery list. The second property, called `middleware`, contains the `async` function that will handle such requests.
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: Everything => All

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 5]
```diff
@@ -266,20 +265,20 @@ import {SINGLETON as UserDAO} from "./user.dao";
...
 
-This generator acts as a last resource in a request, since it does not call `yield next;` like the last one that we've built, simply retrieving an user object, base on the user's e-mail, and then update this user with the new items sent within the `this.request.body`.
+This function acts as a last resource in a request, since it does not call `next();` like the last one that we've built, simply retrieving an user object, base on the user's e-mail, and then update this user with the new items sent within the `ctx.request.body`.
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: last one => previous one

I would end the sentence at "we've built."  Then say: "It simply retrieves a user object, based on the users email, and updates this user with..."

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 6]
```diff
@@ -266,20 +265,20 @@ import {SINGLETON as UserDAO} from "./user.dao";
...
 
-Notice that we use two different objects to get data from the request. The second object, `this.request.body`, contains the new information sent by the user and it is under `koa-bodyparser`'s responsibility to parse it properly. The first object, `this.state.user` does not exist yet. We still have to find a way to be able to identify whose request we are handling, but first we have to wire our new route to our application:
+Notice that we use two different objects to get data from the request. The second object, `ctx.request.body`, contains the new information sent by the user and it is under `koa-bodyparser`'s responsibility to parse it properly. The first object, `ctx.state.user` does not exist yet. We still have to find a way to be able to identify whose request we are handling, but first we have to wire our new route to our application:
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: to be able to => to

no need for the extra "to be able"

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 7]
```diff
@@ -343,27 +342,27 @@ export const SIGN_IN = {
...
 
-Both middlewares described above respond to the user in the same way. If they are fed with proper data they send back a JSON response containing a token - issued by a function called `sign` of [jsonwebtoken package](https://github.com/auth0/node-jsonwebtoken) - and the user data, which contains its e-mail address, its name and its list of items to buy at the grocery store. In case the user sends improper data, like a wrong combination of e-mail and password, or try to register with an e-mail that is already registered, both middlewares answer with an exception describing the problem.
+Both middlewares described above respond to the user in the same way. If they are fed with proper data they send back a JSON response containing a token—issued by a function called `sign` of [jsonwebtoken package](https://github.com/auth0/node-jsonwebtoken)—and the user data, which contains its e-mail address, its name and its list of items to buy at the grocery store. In case the user sends improper data, like a wrong combination of e-mail and password, or try to register with an e-mail that is already registered, both middlewares answer with an exception describing the problem.
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: `sign` of => `sign` of the

try => tries

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 8]
```diff
@@ -845,53 +848,73 @@ And then finally we have reached the point where we have a fully functional groc
 
 ## Aside: Softening the Authentication Burden with Auth0
 
-Creating the authentication mechanisnm was not the hardest task but, for every single application that we build, we will have to recreate it or reuse existing components: one for the front-end application, that will show sign in and sign up forms; and one to handle identity persistence and retrieval.
+Creating the authentication mechanisnm was not the hardest task but, for every single application that we build, we will have to recreate or reuse at least two components: one for the front-end application, that will show sign in and sign up forms; and one to handle identity persistence and retrieval.
 
-Further more, if we want to support identity providers like Google, Facebook, GitHub, etc, then our task will start to become harder. But fear not, [Auth0](https://auth0.com) is here to make our lives easier and securer.
+Further more, if we want to support identity providers (like Google, Facebook, GitHub, etc), multifactor authentication, and so on, then our task will start to become harder. But fear not, [Auth0](https://auth0.com) is here to make our lives easier and securer.
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: securer => more secure

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 9]
```diff
@@ -1032,68 +1057,174 @@ import {SignUpComponent} from "./sign-up/sign-up.component";
...
+<section>
+    <router-outlet></router-outlet>
+</section>
+{% endhighlight %}
 
-We have also added a new `showSignInScreen` method that is responsible for opening the sign in/up screen from Lock. This method will be used later.
+> **Important**—replace `{YOUR-AUTH0-DOMAIN}` above by your Auth0 domain (something like `bkrebs.auth0.com`) and `{YOUR-CLIENT-ID}`, with the *Client Id* of the *Grocery List API (Test Client)* client that was automatically created by Auth0.
 
-We must now update the grocery list component to use the new route that we have created on our backend. We will make three changes to this component:
+When a user successfully signs in to Lock, Auth0 will redirect them to the `http://localhost:3000/callback` URL, which was defined as the `redirect_uri` parameter in the link above.. So, we must create a new Angular component to handle this URL. The following code exhibits the content of the component that we must create in the `src/client/app/callback.component.ts` file:
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: two periods after word "above"

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown [hunk 10]
```diff
@@ -1142,40 +1273,20 @@ import {AuthenticationService} from "./authentication.service";
...
-            <li><a (click)="signIn()">Sign In</a></li>
-        </ul>
-    </div>
-</nav>
-<section>
-    <router-outlet></router-outlet>
-</section>
-```
-
 We are now ready to run our Grocery List application with Auth0 identity management. By issuing `npm run dev` command we shall be able to use it, accessing it on [http://localhost:3000/](http://localhost:3000/), and sign in with Google or any other e-mail address as before.
```
Comment (kukicado) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: I would add the mention that if this is their first time logging in and if they are on localhost they will see the consent dialog.

File: _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown (hunk 3 above)
Comment (brunokrebs) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: Are you sure? This refers to the middleware (singular) not to contents.

Comment (brunokrebs) [inline _posts/2017-01-19-building-and-securing-a-koa-and-angular2-app-with-jwt.markdown]: Sorry, you mean't `this errors`, which makes sense 😄

**Pull request merged**